src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java

```diff
@@ -0,0 +1,444 @@
+      content.add(line);
+    }
+    reader.close();
+
+    return content;
+  }
+
+  private static void writeFile(Path file, List<String> content) throws IOException {
+    BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(file.getOutputStream()));
+    for (String line : content) {
+      writer.write(line);
+      writer.write("\n");
+    }
+    writer.close();
+  }
+
+  private static void applyPatchToFile(List<String> patchContent, Path oldFile, Path newFile,
+      boolean isRenaming)
+      throws IOException, PatchFailedException {
+    // Neither of newFile and oldFile can be none if we should do a rename.
+    if (isRenaming && (oldFile == null || newFile == null)) {
+      throw new PatchFailedException("Cannot rename file without two valid file names");
+    }
+
+    // The file we should read oldContent from.
+    Path inputFile = null;
+    if (oldFile != null && oldFile.exists()) {
+      inputFile = oldFile;
+    } else if (newFile != null && newFile.exists()) {
+      inputFile = newFile;
+    }
+
+    List<String> oldContent;
+    if (inputFile == null)  {
+      oldContent = new ArrayList<>();
+    } else {
+      oldContent = readFile(inputFile);
+    }
+
+    Patch<String> patch = DiffUtils.parseUnifiedDiff(patchContent);
+    List<String> newContent = OffsetPatch.applyTo(patch, oldContent);
+
+    // The file we should write newContent to.
+    Path outputFile;
+    if (oldFile != null && oldFile.exists() && !isRenaming) {
+      outputFile = oldFile;
+    } else {
+      outputFile = newFile;
+    }
+
+    // The old file should always change, therefore we can just delete the original file.
+    // If the output file name is the same as the old file, we'll just recreate it later.
+    if (oldFile != null) {
+      oldFile.delete();
+    }
+
+    // Does this patch look like deleting a file.
+    boolean isDelete = newFile == null && newContent.isEmpty();
+
+    if (outputFile != null && !isDelete) {
+      writeFile(outputFile, newContent);
+    }
+  }
+
+  /**
+   * Strip a number of leading components from a path
+   * @param path the original path
+   * @param strip the number of leading components to strip
+   * @return The stripped path
+   */
+  private static String stripPath(String path, int strip) {
+    int pos = 0;
+    while (pos < path.length() && strip > 0) {
+      if (path.charAt(pos) == '/') {
+        strip--;
+      }
+      pos++;
+    }
+    return path.substring(pos);
+  }
+
+  private static Path getFilePath(String line, int strip, Path outputDirectory) {
+    // The line could look like:
+    // --- a/foo/bar.txt   2019-05-27 17:19:37.054593200 +0200
+    // +++ b/foo/bar.txt   2019-05-27 17:19:37.054593200 +0200
+    // If strip is 1, we want get the file path as <outputDirectory>/foo/bar.txt
+    String file = null;
+    line = line.split("\t")[0];
+    if (line.length() > 4) {
+      if (line.substring(4).trim().equals("/dev/null")) {
```

Comment:
Should we handle "/dev/null" outside this method? I guess this means "does not matter, ignore, because the file is getting added/deleted" - but with returning null from this method, we are mixing it with the case when the file path is just a directory which got stripped? (also, what if strip prefix is too long? - we should specifically indicate this)

Reply:
You're right! We shouldn't mix those cases, I think we can throw an error when failing to calculate the file name. See https://github.com/bazelbuild/bazel/pull/8773